### ISSUE TITLE
net.ltgt.errorprone 0.0.8 not compatible with dagger 2.8

### ISSUE THREAD

**Issue by koral--** [google/error-prone#482]
Not sure if it is bug in errorprone or android/errorprone gradle plugin but first reporting here. Combination mentioned in title causes `NoSuchMethodError: com.google.common.base.Preconditions.checkState` because `errorprone` configuration still uses guava 19.0.

More details: https://github.com/google/dagger/issues/511#issuecomment-262435064

**Comment by tbroyer** [comment]
Note: the `net.ltgt.errorprone` gradle plugin issue tracker is at https://github.com/tbroyer/gradle-errorprone-plugin; yet I don't think this is a bug with the plugin specifically, but rather with ErrorProne itself; possibly due to [this code in JavaC](https://github.com/google/error-prone-javac/blob/90149b22e202e7b24bfafc80df7509c8fb490ee7/src/jdk.compiler/share/classes/com/sun/tools/javac/util/BaseFileManager.java#L133) (called via [this code](https://github.com/google/error-prone-javac/blob/90149b22e202e7b24bfafc80df7509c8fb490ee7/src/jdk.compiler/share/classes/com/sun/tools/javac/processing/JavacProcessingEnvironment.java#L232)), therefore leaking ErrorProne's own class-loader to the annotation processors' class-loader.
Should ErrorProne set the `procloader` option when delegating to JavaC?

**Comment by cushon** [comment]
Error Prone should already be installing a processor classloader that avoids leaking the internal version of guava: https://github.com/google/error-prone/blob/390d90ea894e8270f1c9ddc09df20a01feb1647a/check_api/src/main/java/com/google/errorprone/BaseErrorProneCompiler.java#L176

Either that's broken, or there's something else going on here.

**Comment by koral--** [comment]
Here is example of project which reproduces this issue: https://github.com/koral--/GithubClient
To reproduce run eg. `./gradlew assembleDebug`.

**Comment by tbroyer** [comment]
@cushon The `MaskedClassLoader` still delegates to its parent. In the case of a Gradle build, the `URLClassLoader` returned from `MaskedClassLoader#preRegisterFileManager` internals delegates to its parent (the `MaskedClassLoader`), which delegates to its parent (the `SelfFirstClassLoader` from the `net.ltgt.errorprone` gradle plugin: https://github.com/tbroyer/gradle-errorprone-plugin/blob/v0.0.8/src/main/java/net/ltgt/gradle/errorprone/ErrorProneCompiler.java#L76) which corresponds to the `errorprone` Gradle configuration (a single dependency on `com.google.errorprone:error_prone_core`, which depends on `com.google.errorprone:javac` that brings the `JavaFileManager` used to initialized the `MaskedClassLoader` parent class-loader, which will actually be the same class-loader as the one `MaskedClassLoader` and the rest of Error Prone are loaded from, so the `MaskedClassLoader` is effectively bypassed) and thus contains Guava 19.

Should the gradle plugin strategy be changed? or should `MaskedClassLoader` override `loadClass` et al. instead? (only exposing `java.`, `javax.`, `com.google.errorprone.` and `org.checkerframework.dataflow.`, possibly `sun.` and `com.sun.` too)

**Comment by tbroyer** [comment]
BTW, I can reproduce with Maven too
[test-1.0-SNAPSHOT-project.tar.gz](https://github.com/google/error-prone/files/615774/test-1.0-SNAPSHOT-project.tar.gz)

**Comment by tbroyer** [comment]
Here's the repro-case with Maven, Gradle, and Bazel builds: https://github.com/tbroyer/error-prone-issue-482
Bazel builds OK, Maven and Gradle exhibit the problem.

**Comment by tbroyer** [comment]
FWIW, @koral-- a workaround is to put Guava 20.0 in the Error Prone classpath.
For Gradle:
```gradle
// Override Guava to workaround issue google/error-prone#482
configurations.errorprone.resolutionStrategy.force 'com.google.guava:guava:20.0'
// Alternatively:
/*
dependencies {
  // This is the default value if not set:
  errorprone 'com.google.errorprone:error_prone_core:latest.release'
  // Override Guava to workaround issue google/error-prone#482
  errorprone 'com.google.guava:guava:20.0'
}
*/
```
For Maven:
```
<build>
  <plugins>
    <plugin>
      <groupId>org.apache.maven.plugins</groupId>
      <artifactId>maven-compiler-plugin</artifactId>
      <version>3.6.0</version>
      <configuration>
        <compilerId>javac-with-errorprone</compilerId>
        <forceJavacCompilerUse>true</forceJavacCompilerUse>
        <!-- maven-compiler-plugin defaults to targeting Java 5, but our javac
             only supports >=6 -->
        <source>7</source>
        <target>7</target>
      </configuration>
      <dependencies>
        <dependency>
          <groupId>org.codehaus.plexus</groupId>
          <artifactId>plexus-compiler-javac-errorprone</artifactId>
          <version>2.8.1</version>
        </dependency>
        <!-- override plexus-compiler-javac-errorprone's dependency on
             Error Prone with the latest version -->
        <dependency>
          <groupId>com.google.errorprone</groupId>
          <artifactId>error_prone_core</artifactId>
          <version>2.0.14</version>
        </dependency>
        <!-- Override Guava to workaround issue google/error-prone#482 -->
        <dependency>
          <groupId>com.google.guava</groupId>
          <artifactId>guava</artifactId>
          <version>20.0</version>
        </dependency>
      </dependencies>
    </plugin>
  </plugins>
</build>
```
For Ant, I suppose one can add Guava before `error_prone_ant` in the `<compilerclasspath>` (no idea when simply dropping the error_prone_ant.jar in `$ANT_HOME/lib` or `~/.ant/lib`, are JARs listed in alphabetical order?)

**Comment by koral--** [comment]
@tbroyer it does not seem to work: https://jitpack.io/com/github/koral--/GithubClient/37a6383f6f/build.log

**Comment by tbroyer** [comment]
Ouch! `error_prone_core` actually bundles its dependencies (but also declares them as runtime dependencies), including Guava 19.0, so the Guava 20.0 has to be forced to appear before Error Prone in the classpath; and because [Gradle's `dependencies{}` doesn't give any guarantee on the ordering](https://discuss.gradle.org/t/how-do-i-control-the-order-of-compile-classpaths/5329/4), and `net.ltgt.errorprone` uses Gradle's `Configuration` type rather than `FileCollection` (where all it actually needs is an `Iterable<File>`), I think you were right that there's no workaround for Gradle builds.

I've updated my comment above for Maven and Ant.

**Comment by cushon** [comment]
I was forgetting that the `MaskedClassLoader` approach works for Bazel because javac (but not Error Prone) is on the bootclasspath, so `JavacFileManager.class.getClassLoader()` returns the system classloader. Also that the maven plugin does its own [hacky classloading](https://github.com/sonatype/plexus-compiler/blob/56eac6a4bceb437e6c4bce6e84a1771285915f37/plexus-compilers/plexus-compiler-javac-errorprone/src/main/java/org/codehaus/plexus/compiler/javac/errorprone/JavacCompilerWithErrorProne.java#L99).

So I agree this is broken, but I'm not sure what the best fix is yet.

And we may as well start using guava 20.

**Comment by tbroyer** [comment]
May I suggest starting by removing the maven-shade-plugin from error_prone_core (keep it only in error_prone_ant) so that workarounds are easier should that happen again with Guava 21; and if course updating to Guava 20?

Then we can look at fixing the classloading issue.

**Comment by davido** [comment]
@tbroyer Thanks for the pointer.

 > May I suggest starting by removing the maven-shade-plugin from error_prone_core (keep it only in error_prone_ant)

+1. It is very confusing that maven-shade-plugin is used in artifact with name "core". Moreover, other dependencies are not included in core artifact, like dataflow. It's also unclear to me, why JSR305 also included in core error prone artifact? Is it because of auto value dependency? I understand that auto value simplifies the life of developers. OTOH it poisons your class path and in retrospective, it seems too be a  bad idea and counter productive to use it in low level libraries, that would need to provide end exhibit their dependencies.

FWIW given this class path collision in Bazel JUnit test integration library with auto value version used in Gerrit Code Review, auto value usage in Bazel JUnit test integration was abandoned: https://github.com/bazelbuild/bazel/issues/2044.

Bottom line: error prone core artifact shouldn't provide anything except error prone own classes. Shaded artifact should be called `error_prone_all_in_one` or `error_prone_ant` or some such.

**Comment by tbroyer** [comment]
FWIW, AutoValue has, by design, [zero runtime dependency](https://github.com/google/auto/blob/master/value/userguide/why.md), so it shouldn't _“poison your class path”_, and the bugs in both ErrorProne and Bazel's TestRunner are that AutoValue is bundled in the uberjars.

**Comment by tbroyer** [comment]
I believe this issue can be closed.
